sdk/program/src/loader_upgradeable_instruction.rs

```diff
@@ -0,0 +1,75 @@
+//! Instruction types
+
+use crate::pubkey::Pubkey;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum UpgradeableLoaderInstruction {
+    /// Write program data into a source account.
+    ///
+    /// A source account is an intermediary that once fully populated is used
+    /// with the `Finalize` instruction to populate the program's code account.
+    ///
+    /// # Account references
+    ///   0. [writable, signer] source account to write program data to.
+    Write {
+        /// Offset at which to write the given bytes
+        offset: u32,
+
+        /// Serialized program data
+        #[serde(with = "serde_bytes")]
+        bytes: Vec<u8>,
+    },
+
+    /// Deploy an executable program.
+    ///
+    /// A program consists of an executable account and code account pair.
+    ///   - The executable account's address will serve as the program id in
+    ///     instructions that execute this program.
+    ///   - The code account will remain mutable by the loader only and holds
+    ///     the program data and authority information.
+    ///
+    /// The `Deploy` instruction requires no signers and MUST be included within
+    /// the same Transaction as the system program's `CreateAccount` instruction
+    /// that creates the executable and code accounts. Otherwise another party
+    /// can acquire ownership of the uninitialized accounts.
+    ///
+    /// # Account references
+    ///   0. [writable] The code account.
+    ///   1. [writable] The executable account.
+    ///   2. [writable] The source account where the program data has been
+    ///      written.
+    ///   3. [] The program's authority.
+    ///   4. [] Rent sysvar
+    Deploy,
+
+    /// Upgrade a program.  
+    ///
+    /// A program can be updated as long as the program's authority has not been
+    /// set to `None`.
+    ///
+    /// The source account must contain sufficient lamports to fund the code
+    /// account to be rent-exempt, any additional lamports left over will be
+    /// transferred to the spills account leaving the source account balance at
+    /// zero.
+    ///
+    /// # Account references
+    ///   0. [writable] The code account.
+    ///   1. [] The executable account.
+    ///   2. [] The source account where the program data has been written.
+    ///   4. [signer] The program's authority.
+    ///   5. [writable] The spills account.
+    ///   6. [] Rent sysvar
+    ///   7. [] Slot sysvar
+    Upgrade,
+
+    /// Sets a new program authority that is allowed to update the program.  To
+    /// permanently disable program updates set the authority to `None`.
+    ///
+    /// # Account references
+    ///   0. `[writable]` The code account to change the authority of.
+    ///   1. `[signer]` The current authority.
```

Comment:
IMO pass the new authority as the second key.  None = no second key.
Less custom serde work for non-Rust clients

Reply:
Yeah, much better